pkg/machine/hyperv/vsock.go

```diff
@@ -0,0 +1,270 @@
+	_, err := openVSockRegistryEntry(hv.fqPath())
+	if err == nil {
+		return true, err
+	}
+	if errors.Is(err, registry.ErrNotExist) {
+		return false, nil
+	}
+	return false, err
+}
+
+// findOpenHVSockPort looks for an open random port. it verifies the port is not
+// already being used by another hvsock in the Windows registry.
+func findOpenHVSockPort() (uint64, error) {
+	// If we cannot find a free port in 10 attempts, something is wrong
+	for i := 0; i < 10; i++ {
+		port, err := utils.GetRandomPort()
+		if err != nil {
+			return 0, err
+		}
+		// Try and load registry entries by port to see if they exist
+		_, err = LoadHVSockRegistryEntry(uint64(port))
+		if err == nil {
+			// the port is no good, it is being used; try again
+			logrus.Errorf("port %d is already used for hvsock", port)
+			continue
+		}
+		if errors.Is(err, registry.ErrNotExist) {
+			// the port is good to go
+			return uint64(port), nil
+		}
+		if err != nil {
+			// something went wrong
+			return 0, err
+		}
+	}
+	return 0, errors.New("unable to find a free port for hvsock use")
+}
+
+// NewHVSockRegistryEntry is a constructor to make a new registry entry in Windows.  After making the new
+// object, you must call the add() method to *actually* add it to the Windows registry.
+func NewHVSockRegistryEntry(machineName string, purpose HVSockPurpose) (*HVSockRegistryEntry, error) {
+	// a so-called wildcard entry ... everything from FACB -> 6D3 is MS special sauce
+	// for a " linux vm".  this first segment is hexi for the hvsock port number
+	//00000400-FACB-11E6-BD58-64006A7986D3
+	port, err := findOpenHVSockPort()
+	if err != nil {
+		return nil, err
+	}
+	r := HVSockRegistryEntry{
+		KeyName:     portToKeyName(port),
+		Purpose:     purpose,
+		Port:        port,
+		MachineName: machineName,
+	}
+	if err := r.Add(); err != nil {
+		return nil, err
+	}
+	return &r, nil
+}
+
+func portToKeyName(port uint64) string {
+	// this could be flattened but given the complexity, I thought it might
+	// be more difficult to read
+	hexi := strings.ToUpper(fmt.Sprintf("%08x", port))
+	return fmt.Sprintf("%s-%s", hexi, LinuxVm)
+}
+
+func LoadHVSockRegistryEntry(port uint64) (*HVSockRegistryEntry, error) {
+	keyName := portToKeyName(port)
+	fqPath := fmt.Sprintf("%s\\%s", VsockRegistryPath, keyName)
+	k, err := openVSockRegistryEntry(fqPath)
+	if err != nil {
+		return nil, err
+	}
+	p, _, err := k.GetStringValue(HvsockPurpose)
+	if err != nil {
+		return nil, err
+	}
+
+	purpose, err := toHVSockPurpose(p)
+	if err != nil {
+		return nil, err
+	}
+
+	machineName, _, err := k.GetStringValue(HvsockMachineName)
+	if err != nil {
+		return nil, err
+	}
+	return &HVSockRegistryEntry{
+		KeyName:     keyName,
+		Purpose:     purpose,
+		Port:        port,
+		MachineName: machineName,
+		Key:         k,
+	}, nil
+}
+
+// Listen s used on the windows side to listen for anything to come
+// over the hvsock as a signal the vm is booted
+func (hv *HVSockRegistryEntry) Listen() error {
```

Comment:
Should the Listen function have some sort of timeout in this instance, or is this not a concern? ^^ re: the call from start()